javascript/ql/src/semmle/javascript/security/dataflow/ClientSideUrlRedirectCustomizations.qll

```diff
@@ -191,6 +191,21 @@ module ClientSideUrlRedirect {
     }
   }
 
+  /**
+   * A write to the location using the [history](https://npmjs.com/package/history) library
+   */
+  class HistoryWriteUrlSink extends ScriptUrlSink {
+    HistoryWriteUrlSink() {
+      this =
+        API::moduleImport("history")
+            .getMember(["createBrowserHistory", "createHashHistory"])
+            .getReturn()
+            .getMember(["push", "replace"])
+            .getACall()
+            .getArgument(0)
```

Comment:
The built-in `history` object enforces same-origin policy on the URL, which also blocks the `javascript:` scheme. Are you sure the `history` package doesn't do the same?

Reply:
It works when I try it.  
Copy paste [this compiled JS](https://cdnjs.cloudflare.com/ajax/libs/history/5.0.0/history.development.js) into a browser console, then run the below: 

```JavaScript
var his = HistoryLibrary.createBrowserHistory();
his.push("javascript:alert(2)")
```

But I was too quick with `createHashHistory`, that one is not vulnerable. 

I'm also adding support for the global variable `HistoryLibrary`, that requires a slight refactorization.